tests/test_sessions.py

```diff
@@ -476,6 +476,86 @@ def test_session_by_id_links(client):
     )
     assert rv.status_code == 200
 
+def test_session_by_id_router_interface_add_delete(client):
+    # Establish session
+    client.post(
+          '/api/user',
+          json={
+              'username': 'arthur@projectreclass.org',
+              'password': 'BaLtH@$0R',
+              'first_name': 'Arthur',
+          },
+    )
+
+    rv = client.post(
+        '/api/toynet/session',
+        json={
+            'toynet_topo_id': 1,
+            'toynet_user_id': 'arthur@projectreclass.org',
+        },
+    ) 
+    assert rv.status_code == 201
+    rv_json = json.loads(rv.data.decode('utf-8'))
+    assert rv_json['running'] == True
+    # Save session id
+    session_id = rv_json['toynet_session_id']
+
+    # Add interface to router
+    rv = client.put(
+        f'/api/toynet/session/{session_id}/create/router/interface',
+        json={
+            'name': 'r1',
+            'ip': '100.100.100.100/10',
+        },
+    )
+    assert rv.status_code == 200
+    rv = client.get(f'/api/toynet/session/{session_id}')
+    rv_json = json.loads(rv.data.decode('utf-8'))
+    assert rv_json['topology'][150:181] == '<intf>100.100.100.100/10</intf>'
+
+    # Delete the interface from the router
+    rv = client.put(
+        f'/api/toynet/session/{session_id}/delete/router/interface',
+        json={
+            'name': 'r1',
+            'ip': '100.100.100.100/10',
+        },
+    )
+    assert rv.status_code == 200
+    rv = client.get(f'/api/toynet/session/{session_id}')
+    rv_json = json.loads(rv.data.decode('utf-8'))
+    assert rv_json['topology'][150:172] == '</router></routerList>'
```

Comment:
Is it possible to explain what `150` and `172` are? I am sure these are magic numbers I created in topo ID 1, but for readers who did not design the topology, it is even more magic.

Reply:
They are the index boundaries that a properly-modified topology would have the new router stored at (or in this case, they show the absence of a router that used to be there and the end of the device list). I considered a simpler substring check (eg: `r1 in rv_json['topology']` or `r1 not in rv_json['topology']`) but that does not catch errors where the device is inserted in the wrong list. Most of our tests are  brittle to a changed seed topology

Reply:
Definitely no need to change logic.
I think just having named constants instead of 150 and 172 or a comment explaining what they are would suffice here to help future developers test the right thing, especially if they change the seed topology they need to know your intentions.